server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java

```diff
@@ -264,11 +257,29 @@ public int getCapacity()
     @Override
     public void close() throws IOException
     {
-      log.info("Firehose closing.");
-      closed = true;
-      eventReceiverFirehoseRegister.unregister(serviceName);
-      if (chatHandlerProvider.isPresent()) {
-        chatHandlerProvider.get().unregister(serviceName);
+      if (!closed) {
+        log.info("Firehose closing.");
+        closed = true;
```

Comment:
Should work like this:

1. set closed to true
2. hasMore will use that to exit early if `poll` returns null and return "false"
3. *but* as long as `poll` keeps returning non-null, `hasMore` will keep returning "true"